### ISSUE TITLE
Some metrics do not work for internal containerd

### ISSUE THREAD

**Issue by xhejtman** [rancher/rke2#1167]
Hello,

I have rke2 v1.20.7+rke2r2. I noticed that some cadvisor metrics are empty with internal containerd. I believe this is due to:
/var/lib/rancher/rke2/agent/etc/crictl.yaml containing:
runtime-endpoint: unix:///run/k3s/containerd/containerd.sock

It should contain:
runtime-endpoint: /run/k3s/containerd/containerd.sock

Is there a way to fix this?

**Comment by brandond** [comment]
That is for the crictl CLI tool, and has nothing to do with the kubelet or cadvisor. Are you providing your own containerd, or using the one included with rke2?

**Comment by xhejtman** [comment]
Most of the cluster is using containerd included with rke2. One node due to GPU uses external containerd. The GPU node returns metrics ok while the other nodes return limited metrics like:
container_fs_reads_total{container="",device="/var/lib/kubelet/pods/fcce79d5-9923-4a79-a57e-5b12056c604a/volumes/kubernetes.io~secret/config",id="/",image="",name="",namespace="",pod=""} 0 1624315449301

**Comment by brandond** [comment]
Are you by any chance customizing the container-runtime-endpoint setting, even on nodes using the packaged containerd? See https://github.com/rancher/rke2/issues/1139

**Comment by xhejtman** [comment]
I do not customize c-r-e. 

However, is it expected, that container, image, name, namespace, pod are empty sometimes? if so there might be a different error not related to rke. If not, there is something wrong with my rke2 cluster.

**Comment by brandond** [comment]
I am not aware of any outstanding issues with missing metrics. Is this affecting any particular type of pod or node? Can you provide the output of `kubectl get nodes -o yaml | grep -F node-args` ?

**Comment by xhejtman** [comment]
I do not think it affects only particular type of pods or nodes.

kubectl get nodes -o yaml --context hdhu-cluster | grep -F node-args
      rke2.io/node-args: '["server","--node-taint","CriticalAddonsOnly=true:NoExecute","--tls-san","control-u1.priv.cerit-sc.cz","--tls-san","control-u.priv.cerit-sc.cz","--disable","rke2-ingress-nginx","--cloud-provider-name","external","--kube-apiserver-arg","encryption-provider-config=/etc/rancher/rke2/encryption-provider-config.json","--profile","cis-1.5"]'
      rke2.io/node-args: '["server","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--node-taint","CriticalAddonsOnly=true:NoExecute","--tls-san","control-u2.priv.cerit-sc.cz","--tls-san","control-u.priv.cerit-sc.cz","--disable","rke2-ingress-nginx","--cloud-provider-name","external","--kube-apiserver-arg","encryption-provider-config=/etc/rancher/rke2/encryption-provider-config.json","--profile","cis-1.5"]'
      rke2.io/node-args: '["server","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--node-taint","CriticalAddonsOnly=true:NoExecute","--tls-san","control-u3.priv.cerit-sc.cz","--tls-san","control-u.priv.cerit-sc.cz","--disable","rke2-ingress-nginx","--cloud-provider-name","external","--kube-apiserver-arg","encryption-provider-config=/etc/rancher/rke2/encryption-provider-config.json","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5","--container-runtime-endpoint","/run/containerd/containerd.sock"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'
      rke2.io/node-args: '["agent","--server","https://control-u1.priv.cerit-sc.cz:9345","--token","********","--cloud-provider-name","external","--node-label","storage=local-ssd","--profile","cis-1.5"]'

**Comment by brandond** [comment]
I can confirm that on a stock installation of RKE2 I do see some metrics come back with empy container/image/name fields, but I'm not sure if this is normal cadvisor behavior or not.

```console
[root@centos01 ~]# kubectl get --raw /api/v1/nodes/centos01.lan.khaus/proxy/metrics/cadvisor | grep -F 'container_fs_reads_total{container=""' | wc -l
42
```

**Comment by amincheloh** [comment]
confirm this problem with RKE2 v1.23.7+rke2r2

**Comment by brandond** [comment]
This is an upstream issue, ref:
* https://github.com/kubernetes/website/issues/30681
* https://github.com/kubernetes/kubernetes/issues/89903

in particular this comment that notes that the fix was merged and then reverted, and therefore remains unresolved:
* https://github.com/kubernetes/website/issues/30681#issuecomment-1120173673

**Comment by rohitsakala** [comment]
@brandond can we fill up the deprecated flag `-containerd` in rke2 always as mentioned in point 2 of https://github.com/kubernetes/kubernetes/issues/89903. Is this the right way to go?

**Comment by brandond** [comment]
as far as I know that does not actually resolve this issue, does it?

**Comment by rohitsakala** [comment]
@brandond In this issue https://github.com/kubernetes/kubernetes/issues/89903, under the heading  *Anything else we need to know?*, Point 2 says that 

```Use deprecated parameter to kubelet: --containerd=/u01/data/run/containerd/containerd.sock

That works because it points cadvisor to real containerd socket, and that value is passed to paramater mentioned above. Kubelet started with that parameter logs warning:
kubelet[6694]: Flag --containerd has been deprecated, This is a cadvisor flag that was mistakenly registered with the Kubelet. Due to legacy concerns, it will follow the standard CLI deprecation timeline before being removed.```

He says it works...I didn't check though....Do you think this can be a viable solution ?

**Comment by brandond** [comment]
I don't believe that fixes **everything**. That only fixes the reporter's original issue regarding the socket path:
> cAdvisor in kubelet can't connect to containerd and does not populate container releated fields

I don't believe that's the problem we're running into. cadvisor **is** successfully connecting to containerd and pulling most of the statistics, as seen by the fact that not all metrics are missing. It's just that there are some disk metrics that cadvisor collected when using dockershim that containerd does not make available.

**Comment by rohitsakala** [comment]
Hello @brandond,

Following is the information I gathered and the results of the experiments done.

The cadvisor only reports fs_inodes_free, fs_inodes_total, fs_usage_bytes and fs_limi_bytes for containerd https://github.com/google/cadvisor/pull/2936.
These values are only reported in rke2 if cadvisor is deployed as standalone. 

Other metrics are not fixed. Didn't find any PR or issue. Based on my experiments, I found out that, rke has all metrics and values populated as mentioned in the JIRA issue.

# Experiment Setup 1 - RKE2 v1.24 (cadvisor from kubelet)
curl -sfL https://get.rke2.io/ | INSTALL_RKE2_VERSION=v1.24.2+rke2r1 sh -
systemctl enable rke2-server.service
systemctl start rke2-server.service 
export KUBECONFIG=/etc/rancher/rke2/rke2.yaml PATH=$PATH:/var/lib/rancher/rke2/bin/

## Checking for cadvisor metrics from kubelet
kubectl proxy --address='0.0.0.0' --disable-filter=true &
Visit http://{ip}:8001/api/v1/nodes/ubuntu-s-4vcpu-8gb-intel-sfo3-01/proxy/metrics/cadvisor on browser

## Analysis
Seems like fields are not filled at all like container name container image etc
container_fs_inodes_free = Absent
container_fs_inodes_total = Absent
container_fs_limit_bytes = Absent
container_fs_usage_bytes = Absent
container_fs_io_time_seconds_total = Absent
container_fs_io_current = Absent 
container_fs_io_time_seconds_total = Absent
container_fs_io_time_weighted_seconds_total = Absent
container_fs_write_seconds_total = Absent
container_fs_writes_merged_total = Absent
container_fs_read_seconds_total = Absent
container_fs_reads_merged_total = Absent
container_fs_sector_reads_total = Absent
container_fs_sector_writes_total = Absent

File name - rke2-kubelet.txt

# Experiment Setup 2 - RKE2 v1.24 (cadvisor v0.44.0 as daemonset)
curl -sfL https://get.rke2.io/ | INSTALL_RKE2_VERSION=v1.24.2+rke2r1 sh -
systemctl enable rke2-server.service
systemctl start rke2-server.service 
export KUBECONFIG=/etc/rancher/rke2/rke2.yaml PATH=$PATH:/var/lib/rancher/rke2/bin/

## Deplopying Cadvsior metrics
VERSION=v0.44.0
git clone https://github.com/google/cadvisor
cd cadvisor
sudo snap install kustomize
cd deploy/kubernetes/base && kustomize edit set image gcr.io/cadvisor/cadvisor:${VERSION} && cd ../../..
kubectl kustomize deploy/kubernetes/base
kubectl kustomize deploy/kubernetes/base | kubectl apply -f -
Create a nodeport service with namespace cadvisor

## Checking for cadvisor metrics from workload as daemonset
Visit http://{ip}:{nodeport}/metrics on browser

## Analysis 
Seems like all fields are filled but they are no fields related to container name, container image etc
container_fs_inodes_free = Present
container_fs_inodes_total = Present
container_fs_limit_bytes = Present
container_fs_usage_bytes = Present

container_fs_io_time_seconds_total = Absent
container_fs_io_current = Absent 
container_fs_io_time_seconds_total = Absent
container_fs_io_time_weighted_seconds_total = Absent
container_fs_write_seconds_total = Absent
container_fs_writes_merged_total = Absent
container_fs_read_seconds_total = Absent
container_fs_reads_merged_total = Absent
container_fs_sector_reads_total = Absent
container_fs_sector_writes_total = Absent

File name - rke2-cadvisor-ds.txt

# Experiment Setup 3 - RKE2 v1.24 (cadvisor:containerd-cri as daemonset)
curl -sfL https://get.rke2.io/ | INSTALL_RKE2_VERSION=v1.24.2+rke2r1 sh -
systemctl enable rke2-server.service
systemctl start rke2-server.service 
export KUBECONFIG=/etc/rancher/rke2/rke2.yaml PATH=$PATH:/var/lib/rancher/rke2/bin/

## Deplopying Cadvsior metrics
VERSION=v0.41.0-containerd-fix-rc.2
git clone https://github.com/google/cadvisor
cd cadvisor
sudo snap install kustomize
cd deploy/kubernetes/base && kustomize edit set image gcr.io/cadvisor/cadvisor:${VERSION} && cd ../../..
kubectl kustomize deploy/kubernetes/base
kubectl kustomize deploy/kubernetes/base | kubectl apply -f -
Create a nodeport service with namespace cadvisor

## Checking for cadvisor metrics from workload as daemonset
Visit http://147.182.255.124/:{nodeport}/metrics on browser

## Analysis 
Seems like all fields are filled but they are no fields related to container name, container image etc
container_fs_inodes_free = Present
container_fs_inodes_total = Present
container_fs_limit_bytes = Present
container_fs_usage_bytes = Present

container_fs_io_time_seconds_total = Absent
container_fs_io_current = Absent 
container_fs_io_time_seconds_total = Absent
container_fs_io_time_weighted_seconds_total = Absent
container_fs_write_seconds_total = Absent
container_fs_writes_merged_total = Absent
container_fs_read_seconds_total = Absent
container_fs_reads_merged_total = Absent
container_fs_sector_reads_total = Absent
container_fs_sector_writes_total = Absent

File name - rke2-cadvisor-ds-cri.txt

# Experiment RKE kubelet Cadvisor
rke up

## Checking for cadvisor metrics from workload as daemonset
kubectl proxy --address='0.0.0.0' --disable-filter=true &
Visit http://{ip}:8001/api/v1/nodes/rohitsakala/proxy/metrics/cadvisor on browser

## Analysis
Works nicely. All metrics populated and container name and images intact.

[rke2-kubelet.txt](https://github.com/rancher/rke2/files/9148600/rke2-kubelet.txt)
[rke-cadvisor-kubelet.txt](https://github.com/rancher/rke2/files/9148596/rke-cadvisor-kubelet.txt)
[rke2-cadvisor-ds-cri.txt](https://github.com/rancher/rke2/files/9148598/rke2-cadvisor-ds-cri.txt)
[rke2-cadvisor-ds.txt](https://github.com/rancher/rke2/files/9148599/rke2-cadvisor-ds.txt)

**Comment by brandond** [comment]
> rke has all metrics and values populated as mentioned in the JIRA issue
> These values are only reported in rke2 if cadvisor is deployed as standalone.

Yes, you've confirmed the issue. The problem is that upstream Kubernetes embeds a **specific** version of cadvisor in each release of the kubelet. RKE2 packages unmodified Kubernetes code; this issue will not be fixed in RKE until upatream updates the embedded cadvisor version to a release that includes the fix.

**Comment by rohitsakala** [comment]
> > rke has all metrics and values populated as mentioned in the JIRA issue
> > These values are only reported in rke2 if cadvisor is deployed as standalone.
> 
> Yes, you've confirmed the issue. The problem is that upstream Kubernetes embeds a **specific** version of cadvisor in each release of the kubelet. RKE2 builds Kubernetes unmodified from upstream; this issue will not be fixed in RKE2 until upstream updates the embedded cadvisor version to a release that includes the fix.

I also confirmed that only 4 out of all missing metrics mentioned in the Jira issue are fixed in the latest version of cadvisor.

**Comment by brandond** [comment]
FYI - it looks like upstream is working this through a KEP (Kubernetes Enhancement Proposal) that will move the stats collection from cadvisor into the CRI interface: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md

I suspect this means that the issue will not be fixed upstream in the near term. If anyone is aware of a forked version of cadvisor or containerd that resolve this issue in the short term, I'd be interested in seeing if we can pull that into RKE2.
